### PR TITLE
feat(sync): hold the checkout lock through hosted ledger reads

### PR DESCRIPTION
**Pull request by shrimalmadhur**
> guided&nbsp;by&nbsp;<picture><source media="(prefers-color-scheme: dark)" srcset="https://sageox.ai/sageox-wordmark-dark.png">SageOx</picture>&nbsp;&middot;&nbsp;<a href="https://sageox.ai/c/ses_01a0a5d8-6775-7246-a2f6-6cd2173cc37e">Session</a>

## What broke

- **The guard was never wired up.** #893 shipped `ledger.WithReadCheckout` — the function that holds the materialization lock across a read — but it is called from **test files only**. No shipped command holds the checkout lock.
- **#891 rules that out explicitly:** *"An internal helper exercised only by tests, or `--check` followed by an unlocked CLI reader, does not satisfy this."*
- **It is what agent-toolkit#57 blocks on:** *"Require a released ox reader command that holds the guard for the complete content read. Do not port lock-path hashing or receipt verification into TypeScript to compensate for an unguarded CLI."*
- **The readers toolkit actually calls** (`packages/core/src/team-server.ts`) are `ox session list` and `ox glance`. Neither could reach a hosted ledger at all — both resolve it from the working directory, and a hosted runtime has no source checkout.

```mermaid
flowchart LR
    R[ox sync --read-only] -->|takes lock| L[(checkout lock)]
    L --> M[invalidate receipt → mutate → verify → publish]
    S[ox session list --repo repo_uuid] -->|takes SAME lock| L
    G[ox glance --repo repo_uuid] -->|takes SAME lock| L
    L --> C[verify receipt vs HEAD + coverage] --> D[read every file, then release]
```

## What this PR ships

- **One guard, `withHostedLedger`** (`cmd/ox/ledger_read.go`) — takes the same lock materialization takes, verifies the receipt against the current HEAD and coverage, runs the read, and only then releases.
- **`ox session list --repo repo_<uuid>`** and **`ox glance --repo repo_<uuid>`** route through it. `--repo` on session list keeps its filesystem-path form; a path cannot spell a canonical `repo_<uuid>`, so the two forms cannot collide. `ox glance --repo` has no path form and rejects one rather than silently answering about the current project.
- **Shared selection.** `selectHostedLedger` is extracted out of `cmd/ox/sync_read.go`, so `ox sync --read-only` and the readers cannot diverge on which endpoint, data home, or repo identity is acceptable. No working directory, project config, source remote, or disk login participates.
- **Collect under the lock, render after it.** A consumer that drains stdout slowly cannot extend how long the refresher is blocked.
- **Failures are sanitized and unmistakable.** `Ledger read failed: <class>` on stderr, **stdout empty**, exit `1` refused / `2` unsafe selection. A refused read can never be read as a ledger that happens to be empty.
- **`ledger.ReadNotReadyError`** so a CLI reader surfaces the guard's verdict without pattern-matching an error string.
- **Docs:** a new *Guarded reader commands* section in `docs/specs/ledger-read-sync.md` plus regenerated reference pages.

## Test Plan

Six new tests in `cmd/ox/ledger_read_test.go`, all driving **shipped commands** against a real ledger served over Git smart HTTP (`git-http-backend` + TLS fixture), materialized by the real `ox sync --read-only` — not a hand-assembled directory.

- `TestHostedGuardHoldsTheLockForTheWholeRead` — a second acquisition of the same lock, from inside the read, must time out
- `TestHostedReaderSerializesAgainstConcurrentRefresh` — a real reader command racing a refresh
- `TestHostedReadersServeTheMaterializedLedger` — sessions, murmurs, `--all`, and `--check` still ready afterwards (no reader writes)
- `TestHostedReadersRefuseUnverifiedCheckouts`, `TestHostedSessionListRendersATable`, `TestHostedLedgerSelection`, `TestOnlySessionListAmongListCommandsTakesARepoFlag`

### Red-first proof

Replacing the guard with the `--check`-then-read shape #891 rules out:

```
--- FAIL: TestHostedGuardHoldsTheLockForTheWholeRead
    Error:    An error is expected but got nil.
    Messages: a refresh must not be able to take the checkout lock while a read is in flight
```

Restored → green.

**This mattered.** My first race test passed *with the guard removed*, because `CheckReadiness` also takes the lock and gives it straight back — it proved only "the reader touches the lock," not "holds it through the read." A gate nobody watched fail is a gate nobody has tested; that one would have shipped green and guarded nothing.

Also verified red: deleting materialized content yields `dirty`, not a smaller session list.

### Gates

`make format` · `make lint` (0 issues) · `make test` (**18,642 passed, 0 failed**) · new tests pass under `-race` · real-binary smoke test of both readers and the unchanged project paths.

## Scope

**Closes one work item on #891 — not the issue.** Still outstanding: #945, #946, #947, #948, the live acceptance criteria, and the release the last box requires.

The reconciliation comment on #891 lists #945–#948 as what remains; it does not cover this item, because the live scale run exercised the sync path and never reached the reader guard. Left a note there so the next person walking the checklist sees both.

**For the parallel #945–#948 work:** this touches `internal/ledger/read_sync.go` only around `WithReadCheckout` — #945 is in `readSyncLocked`, #946/#947/#948 are in `hydrateReadFiles`. No textual overlap. The one real coupling is that the sanitized class vocabulary now has **two** renderers instead of one; detail on #946.

**Ordering note:** readers and materialization share one exclusive lock and `RepoLockTimeout` is 2 minutes, so #948's ~80-minute cold clone means a hosted reader returns `interrupted` for that whole window. Correct per the spec's separate-budgets rule, but it makes #948 the gate on time-to-first-successful-read. Detail on #948.

## Review

**Ryan's review required** — this changes data-access ergonomics (how a hosted consumer reaches ledger content).

- Cached readiness is never permission. These commands read local state only, contact no server, and neither establish nor renew authorization — a consumer must check current repo authorization separately.
- Before the first refresh the class is `interrupted`, matching `--check`: readiness is invalidated *before* mutation, so "no receipt" is indistinguishable from "refresh interrupted before publishing one." Pre-existing; now documented rather than changed.
- Remaining live validation specific to this change: **how long a guarded read holds the lock on a 15.5k-object ledger.** Cheap to measure now that such a ledger exists.

SageOx-Session: https://sageox.ai/c/ses_01a0a5d8-6775-7246-a2f6-6cd2173cc37e

---
<a href="https://app.blacksmith.sh/sageox/codesmith/ox/pr/950"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1792083035&installation_model_id=433550&pr_number=950&repository=sageox%2Fox&return_to=https%3A%2F%2Fgithub.com%2Fsageox%2Fox%2Fpull%2F950&signature=b4fd53aac2f55118ebf82a486f0de19511841a7b173e18514ec7c82be6f89780"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

- **New Features**
  - `ox glance` can read activity from hosted ledgers using a canonical repository ID.
  - `ox session list --repo` can display sessions from hosted ledgers without a local project checkout.
  - Hosted reads support existing filters, time windows, table output, and JSON output.

- **Bug Fixes**
  - Improved validation and clearer failures for invalid repositories, unavailable ledgers, interrupted reads, and unready checkouts.
  - Local checkpoints are preserved when output fails; hosted reads use default windows without advancing checkpoints.
  - Empty results now return well-formed JSON arrays.

- **Documentation**
  - Added hosted-ledger usage examples and documented repository selection and read behavior.

### CODE REVIEW

File: cmd/ox/ledger_read.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+			return ""
+		}
+		name, value, split := strings.Cut(arg, "=")
+		if name != "--repo" {
+			continue
+		}
+		if !split && i+1 < len(args) {
+			value = args[i+1]
+		}
+		return value
```
[the remark this answers is not in the document]
Comment (shrimalmadhur) [reply cmd/ox/ledger_read.go]: Fixed in bb6c2209 — real bug, and the reproducer was decisive.

`hostedLedgerRepoArg` now returns the **last** `--repo`, matching what Cobra binds. Taking the first meant `--repo /local/path --repo repo_<uuid>` ran the project prelude — including dotenv from the working directory — before runtime selected the hosted checkout. That is exactly the working-directory influence this path exists to deny, so it was a real hole and not a corner case.

Regression test: three new rows in `TestHostedLedgerSelection`, covering the repeated flag in both orders and the `--repo=` spelling. Verified red-first — restoring the first-wins loop fails on `[session list --repo /local/path --repo repo_019ff...]`.

File: cmd/ox/ledger_read.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	// stdout that a consumer would have to guess at.
+	ctx, stop := signal.NotifyContext(cmd.Context(), os.Interrupt, syscall.SIGTERM)
+	defer stop()
+
+	var readErr error
+	guardErr := ledger.WithReadCheckout(ctx, path, repoID, ep, func(ledger.ReadSyncResult) error {
+		readErr = read(path)
+		// The reader's own failure is returned below with its real error. Only
+		// the guard's verdict travels through this return value.
+		return nil
```
[the remark this answers is not in the document]
Comment (shrimalmadhur) [reply cmd/ox/ledger_read.go:116]: Fixed in bb6c2209.

You and CodeRabbit found the same hole from different angles. `withHostedLedger` now re-checks `ctx.Err()` after the guard returns and maps it to the documented `interrupted` class before anything is rendered. The bug was worse than a missing check: the comment directly above claimed the signal "turns that into a sanitized failure," which the code did not do.

On prompt cancellation of the traversal itself: the ledger readers reach `session.Store.ListSessions` and `glance.Harvest*`, which are context-free across two packages, so plumbing cancellation into them is a real refactor rather than a guard change. Deliberately not doing that here — the correctness defect you demonstrated (reporting success for an abandoned read) is fixed, and the lock is released by the kernel when the process exits, so the held-lock symptom resolves on its own.

Regression test: `TestHostedReadCanceledMidReadIsNotReportedAsSuccess` cancels while the read is in flight — exactly your scenario. Verified red-first.

File: cmd/ox/glance.go [hunk 3]
```diff
@@ -24,70 +25,126 @@ Examples:
...
 	// Derive repo name from project root
 	projectRoot, _ := requireProjectRoot()
-	repo := glanceRepoName(projectRoot)
+	data, err := glanceActivity(ledgerPath, glanceRepoName(projectRoot), since, until)
+	if err != nil {
+		return err
+	}
+
+	// Advance checkpoint so next invocation without --since starts from now
+	_ = glance.MarkRead(ledgerPath)
```
[the remark this answers is not in the document]
Comment (shrimalmadhur) [reply cmd/ox/glance.go:98]: Fixed in bb6c2209 — and this one was a real data-loss bug, not just ordering.

The write now happens first; `MarkRead` runs only after it succeeds. Advancing first meant a closed pipe or a failed write skipped that window **permanently**: the next bare `ox glance` resumes from a checkpoint past activity the consumer never received. Silent under-reporting, which is the worst direction for this command to fail in.

Worth noting the ordering predates this PR — I moved the code into `glanceActivity` and carried it along, so the diff surfaced it rather than introduced it.

Regression test: `TestProjectGlanceKeepsTheCheckpointWhenOutputFails` writes to a closed `io.Pipe` and asserts the checkpoint still resolves to the fallback window rather than ~now. Verified red-first — with the old ordering it fails with `"343µs" is not greater than "1h0m0s"`.

File: cmd/ox/ledger_read_test.go [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+			require.NoError(t, os.RemoveAll(dir))
+			close(held)
+			<-release
+			// Put the ledger back before the reader is allowed to look at it.
+			require.NoError(t, os.MkdirAll(dir, 0700))
+			for name, body := range saved {
+				require.NoError(t, os.WriteFile(filepath.Join(dir, name), body, 0600))
+			}
+			return nil
+		}))
```
[the remark this answers is not in the document]
Comment (shrimalmadhur) [reply cmd/ox/ledger_read_test.go]: Fixed in bb6c2209. Agreed on the mechanism and the shape of the fix.

`require.*` in a worker goroutine calls `runtime.Goexit`, so `close(held)` never runs and the test **hangs** instead of failing — which in CI means burning to the 10-minute timeout rather than reporting a failure. Applied as suggested: `assert.*` plus an explicit return after each check inside the callback, a separate `refreshDone` signal, `select` on both before starting the reader, and a non-blocking receive on `listed` so a reader that failed cannot block the final assertion. No deferred `close(held)`, so the midpoint signal keeps its meaning.

File: cmd/ox/ledger_read.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	// A hosted runtime bounds the tool call by killing the process. Catching
+	// the signal turns that into a sanitized failure instead of a truncated
+	// stdout that a consumer would have to guess at.
+	ctx, stop := signal.NotifyContext(cmd.Context(), os.Interrupt, syscall.SIGTERM)
+	defer stop()
+
+	var readErr error
+	guardErr := ledger.WithReadCheckout(ctx, path, repoID, ep, func(ledger.ReadSyncResult) error {
+		readErr = read(path)
```
[the remark this answers is not in the document]
Comment (shrimalmadhur) [reply cmd/ox/ledger_read.go:113]: Fixed in bb6c2209 — see the fuller reply on the Greptile thread covering the same defect.

Short version: `withHostedLedger` now re-checks `ctx.Err()` after the guard returns and maps it to `interrupted` before anything renders, so a signal arriving mid-read can no longer surface as a successful result. Not plumbing cancellation into `glance.Harvest*` / `session.Store.ListSessions` here — those are context-free across two packages and that is a refactor of its own; the reporting defect is what made this unsafe.

Regression test: `TestHostedReadCanceledMidReadIsNotReportedAsSuccess`, verified red-first.

File: docs/reference/session/list.mdx [hunk 6]
```diff
@@ -25,11 +25,18 @@ session data reflects the --repo target, but ledger-merged sessions still
 come from the CURRENT directory's configured ledger — this only matters
 if the two repos use different team ledgers.
 
+Pass a canonical repo_<uuid> to --repo instead to read a hosted ledger kept
+by 'ox sync --read-only'. That form needs no checkout and no project: it
+selects the ledger with SAGEOX_ENDPOINT and XDG_DATA_HOME, and reads it
+while holding the same lock the refresh takes, so a concurrent refresh
+cannot show a half-written checkout. See docs/specs/ledger-read-sync.md.
```
[the remark this answers is not in the document]
Comment (shrimalmadhur) [reply docs/reference/session/list.mdx]: Fixed in bb6c2209, with two things worth recording.

**The fix needed the concatenation form.** `sessionListCmd.Long` is a backtick-delimited raw string literal, so adding literal backticks terminated it and broke the build. The repo hit this before in `ca67d780` and solved it with `` ` + "`repo_<uuid>`" + ` ``; I matched that. `docs/reference/session/list.mdx` and `glance.mdx` are now clean when inline-code spans are stripped, which is the same rule `TestPRHeaderLongHasNoBareHTMLTags` applies.

**The finding is real but this fix does not resolve it.** Scanning every generated page for unfenced `<tag>` outside inline code turns up **63 occurrences across 26 files** — `<name>`, `<path>`, `<uuidv7>`, `<event>`, `<subject>` and others — all pre-existing. Mine was one of them. So either MDX compilation is already failing for `@sageox/cli-docs` and has been for a while, or the downstream compiler tolerates unknown lowercase tags and the hazard is narrower than the guard test assumes. Both readings are worth resolving, and neither is answerable inside this PR.

Two follow-ups that belong in their own change: generalize `TestPRHeaderLongHasNoBareHTMLTags` to walk the whole command tree (it fails on 20+ commands today), and sweep the remaining placeholders. Flagging rather than doing a 63-occurrence refactor in a PR about the checkout lock.

**Comment by shrimalmadhur** [comment]
@greptileai

File: cmd/ox/ledger_read.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+//
+// The LAST occurrence wins, because that is the one Cobra binds. Returning the
+// first would let `--repo /some/path --repo repo_<uuid>` run the project
+// prelude — loading dotenv from the working directory — and only then select
+// the hosted checkout, which is exactly the influence this path exists to deny.
+func hostedLedgerRepoArg(args []string) string {
+	repo := ""
+	for i, arg := range args {
+		if arg == "--" {
+			break
```
[the remark this answers is not in the document]
Comment (shrimalmadhur) [reply cmd/ox/ledger_read.go:41]: Fixed in 72651459. Verified before trusting it — this reproduces exactly as described:

```
argv=[--repo -- --repo repo_019ff...]  ->  cobra binds repo="repo_019ff..." ; scanner returned "--"
```

pflag hands `--` to the first `--repo` as its value, so reading it as the positional separator abandoned the scan and left `headlessLedgerReadRequested` false while Cobra bound the hosted repo. The prelude would then load `.env.local` / `.env` from the working directory before the hosted handler ran, which breaks the isolation contract. Applied the index-stepping fix as suggested.

Regression rows added to `TestHostedLedgerSelection` for both `--repo -- --repo X` and `--repo=-- --repo X`, verified red-first — reverting to `args[i+1]` fails on `[session list --repo -- --repo repo_019ff...]`.

Worth recording for whoever reviews this next: that is the **third** defect in the same surface — preflight hand-scanning argv and disagreeing with what pflag actually binds (Greptile caught the repeated-flag case, you caught this one). A hand-rolled scanner re-implements a parser badly by construction. The structurally correct fix is for preflight to parse with a real `pflag.FlagSet` carrying the same flag definitions instead of scanning tokens. That is a bigger change than this PR should absorb, but if a fourth shape turns up, that is the fix rather than another special case.

**Comment by shrimalmadhur** [comment]
@greptileai

**Comment by shrimalmadhur** [comment]
@coderabbitai review
